src/sqlfluff/rules/L059.py

```diff
@@ -6,36 +6,118 @@
-    | Use unquoted identifiers where possible.
+    | If ``force_quote_identifier = False`` and ``preferred_quote_identifier = '"'`` (or unspecified, as these are the default)
+    | Use unquoted identifiers where possible, and use '"' whevener necessary.
+
+    .. code-block:: sql
+
+        SELECT 123 as foo, "sum"
 
+    | If ``force_quote_identifier = True`` and ``preferred_quote_identifier = "`"``
+    | Use preferred quote identifiers any time.
+    
     .. code-block:: sql
 
-        SELECT 123 as foo
+        SELECT 123 as `foo`, `sum`
 
     """
 
+    config_keywords = [
+        "force_quote_identifier",
+        "preferred_quote_identifier"
+    ]
+    
     def _eval(self, context: RuleContext) -> Optional[LintResult]:
         """Unnecessary quoted identifier."""
-        # We only care about quoted identifiers.
-        if context.segment.name != "quoted_identifier":
+        # Config type hints
+        self.force_quote_identifier: bool
+        self.preferred_quote_identifier: str
+        
+        quote = self.preferred_quote_identifier
+        
+        # If the configuration says to not force the identifier quoting, then
+        # simply care about the quoted identifiers.
+        context_policy = ("quoted_identifier")
+        if self.force_quote_identifier == True:
+            context_policy = ("naked_identifier", "quoted_identifier")
+        
+        # We only care about quoted identifiers and possibly naked identifiers if 
+        # the configuration setting is forcing identifiers to be quoted.
+        if context.segment.name not in context_policy:
             return None
-
+        
+        # Manage simplest case of quoted identifiers must be forced first.
+        if self.force_quote_identifier == True:
+            if context.segment.name == "naked_identifier":
+                identifier_contents = context.segment.raw
+            else:
+                identifier_contents = context.segment.raw[1:-1]
+                
+            # If the identifier_contents contains any occurrence of the preferred quote,
+            # ignore this rule for now.
+            if quote in identifier_contents:
+                # TODO: if different levels of criticity can be set, this case should
+                #       trigger an `info` level.
+                return None
+            
+            expected = f"{quote}{identifier_contents}{quote}"
+            
+            # For naked identifiers, make the rule fails directly.
+            if context.segment.name == "naked_identifier":
+                return LintResult(
+                    context.segment,
+                    fixes=[
+                        LintFix.replace(
+                            context.segment,
+                            [CodeSegment(raw=expected, name="quoted_identifier", type="identifier")],
+                        )
+                    ],
+                    description=f"Missing quoted identifier {expected}.",
+                )
```

Comment:
So we're barring `naked_identifiers`? And forcing even them to be quoted?

That feels like a separate option to what I thought this was.

Should we split this into:
- `force_quote_identifiers` - replaces unquoted identifiers with quoted version (e..g. `foo` to `"foo"`)
- `force_quote_type` - Changes quotes to preferred type (e.g. `'foo'` ro `"foo"`)

Some might want the second without the first.

Reply:
I don't know very well the segment categorizations. I thought `naked_identifiers` is a synonym for "unquoted identifier". This isn't the case?

My point of this feature is to actually have a configuration option to replaces unquoted identifiers (even if unnecessary) with their quoted version (using the quote defined into `preferred_quote_identifier`), but also changes quotes to preferred type using `preferred_quote_identifier` config.

If some want the second without the first, using `force_quote_identifier = False` is doing the job.

Reply:
> I don't know very well the segment categorizations. I thought naked_identifiers is a synonym for "unquoted identifier". This isn't the case?

That is the case.

> My point of this feature is to actually have a configuration option to replaces unquoted identifiers (even if unnecessary) with their quoted version

Yes I got that, when reading the code. Was a bit shocked to be honest, but hey if that's what you want! 😁

But we need a way of preserving the existing L059 functionality (remove unnecessary quotes, but don't add them). Which I don't think your current config supports does it? To me that should be the default, to be backwards compatible, and any of your new asks, should be configurable extras for those that want it. Maybe that could be done by having the `force_identifiers_quote_type` to be not set? So it's ignored?

Reply:
> we need a way of preserving the existing L059 functionality (remove unnecessary quotes, but don't add them).

I was trying to do this actually. With the default configuration I provided (`force_quote_identifier = False`), then unnecessary quotes must be removed, and the L059 rule fails if they're not.
My main issue was to find the expected quote for identifiers when `force_quote_identifier = True` (or when the identifier is a reserved keyword/has special chars), which is why I added the configuration `preferred_quote_identifier` so that each user can define its own.

Reply:
_Resolved as wanted behavior has changed_